lib/Dialect/FIRRTL/Transforms/ExtractInstances.cpp

```diff
@@ -670,26 +683,53 @@ void ExtractInstancesPass::extractInstances() {
         // was rooted at.
         if (nlaIdx == 0) {
           LLVM_DEBUG(llvm::dbgs() << "    - Re-rooting " << nlaPath[0] << "\n");
+          assert(nlaPath[0].isa<InnerRefAttr>() &&
+                 "head of hierpath must be an InnerRefAttr");
           nlaPath[0] =
               InnerRefAttr::get(newParent.moduleNameAttr(),
                                 nlaPath[0].cast<InnerRefAttr>().getName());
-          auto builder = OpBuilder::atBlockBegin(getOperation().getBody());
 
-          auto newNla = builder.create<HierPathOp>(
-              newInst.getLoc(), circuitNamespace.newName(nla.sym_name()),
-              builder.getArrayAttr(nlaPath));
-          auto nlaAnno = instNonlocalAnnos.find(nla);
-          if (nlaAnno != instNonlocalAnnos.end()) {
-            Annotation newAnno = nlaAnno->getSecond();
-            newAnno.setMember("circt.nonlocal",
-                              FlatSymbolRefAttr::get(newNla.sym_nameAttr()));
-            newInstNonlocalAnnos.push_back(newAnno);
+          if (instParentNode->hasOneUse()) {
+            // Simply update the existing NLA since our parent is only
+            // instantiated once, and we therefore are not creating multiple
+            // instances through the extraction.
+            nla.namepathAttr(builder.getArrayAttr(nlaPath));
+            for (auto anno : instNonlocalAnnos.lookup(nla))
+              newInstNonlocalAnnos.push_back(anno);
+            nlaTable.erase(nla);
+            nlaTable.addNLA(nla);
+            LLVM_DEBUG(llvm::dbgs() << "    - Modified to " << nla << "\n");
+          } else {
+            // Since we are extracting to multiple parent locations, create a
+            // new NLA for each instantiation site.
+            auto builder = OpBuilder::atBlockBegin(getOperation().getBody());
+            auto newNla = builder.create<HierPathOp>(
+                newInst.getLoc(), circuitNamespace.newName(nla.sym_name()),
+                builder.getArrayAttr(nlaPath));
+            for (auto anno : instNonlocalAnnos.lookup(nla)) {
+              anno.setMember("circt.nonlocal",
+                             FlatSymbolRefAttr::get(newNla.sym_nameAttr()));
+              newInstNonlocalAnnos.push_back(anno);
+            }
+
+            nlaTable.erase(nla);
```

Comment:
Can we remove the `nla` here, is it possible that the `nla` is reused by other annotations ?

Reply:
Excellent point! Don't think we can remove it here. Will be fixed in the rebase.